arangosh/Utils/ProgressTracker.h

```diff
@@ -0,0 +1,108 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2020 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Lars Maier
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef ARANGOSH_UTILS_PROGRESS_TRACKER_H
+#define ARANGOSH_UTILS_PROGRESS_TRACKER_H 1
+
+#include "ManagedDirectory.h"
+namespace arangodb {
+template <typename T>
+struct ProgressTracker {
+  ProgressTracker(ManagedDirectory& directory, bool ignoreExisting);
+
+  ProgressTracker(ProgressTracker const&) = delete;
+  ProgressTracker(ProgressTracker&&) noexcept = delete;
+  ProgressTracker& operator=(ProgressTracker const&) = delete;
+  ProgressTracker& operator=(ProgressTracker&&) noexcept = delete;
+
+  T getStatus(std::string const& collectionName);
+  void updateStatus(std::string const& collectionName, T const& status);
+
+  ManagedDirectory& directory;
+  std::shared_mutex _collectionStatesMutex;
+  std::mutex _writeFileMutex;
+  std::unordered_map<std::string, T> _collectionStates;
+  std::atomic<bool> _writeQueued{false};
+};
+
+template <typename T>
+void ProgressTracker<T>::updateStatus(std::string const& collectionName,
+                                                      T const& status) {
+  {
+    std::unique_lock guard(_collectionStatesMutex);
+    _collectionStates[collectionName] = status;
+
+    if (_writeQueued) {
+      return;
+    }
+
+    _writeQueued = true;
+  }
+
+  {
+    std::unique_lock guard(_writeFileMutex);
+    VPackBuilder builder;
+    {
+      std::unique_lock guardState(_collectionStatesMutex);
+      _writeQueued = false;
```

Comment:
in theory it is possible that this line is not executed, e.g. if line 63 throws an out-of-memory exception.
Probably we should wrap this in a try catch.

Reply:
I changed it as follows: the Builder is initialized using a buffer that is declared above. The default constructor of a buffer is `noexcept`.